lib/dialects/abstract/connection-manager.js

```diff
@@ -312,6 +288,36 @@ class ConnectionManager {
     if (!this.dialect.connectionManager.validate) return true;
     return this.dialect.connectionManager.validate(connection);
   }
+
+  _acquire(options) {
+    const connectionPromise = this.pool.acquire(options.priority, options.type, options.useMaster);
+    return this._watchPoolError(connectionPromise).tap(() => { debug('connection acquired'); });
+  }
+
+  _watchPoolError(acquirePromise) {
```

Comment:
@felixfbecker I agree with the general ideology but this promise in some condition will never be resolved, thats why we used to watch it using timers initially, now using recursive watcher method. There is no other way around, we have to pass promise to method which will call itself with same promise